aptos-move/parallel-executor/src/output_delta_resolver.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::delta_change_set::{deserialize, serialize};
+use aptos_types::write_set::{TransactionWrite, WriteOp};
+use mvhashmap::{EntryCell, MVHashMap};
+use std::{hash::Hash, thread::spawn};
+
+/// Resolved and serialized data for WriteOps, None means deletion.
+pub type ResolvedData = Option<Vec<u8>>;
+
+pub struct OutputDeltaResolver<K, V> {
+    versioned_outputs: MVHashMap<K, V>,
+}
+
+impl<K: Hash + Clone + Eq + Send + 'static, V: TransactionWrite + Send + Sync + 'static>
+    OutputDeltaResolver<K, V>
+{
+    pub fn new(versioned_outputs: MVHashMap<K, V>) -> Self {
+        Self { versioned_outputs }
+    }
+
+    /// Takes Self, vector of all involved aggregator keys (each with at least one
+    /// delta to resolve in the output), resolved values from storage for each key,
+    /// and blocksize, and returns a Vec of materialized deltas per transaction index.
+    pub fn resolve(
+        self,
+        aggregator_keys: Vec<(K, anyhow::Result<ResolvedData>)>,
+        block_size: usize,
+    ) -> Vec<Vec<(K, WriteOp)>> {
+        let mut ret: Vec<Vec<(K, WriteOp)>> = (0..block_size).map(|_| Vec::new()).collect();
+
+        // TODO: with more deltas, re-use executor threads and process in parallel.
+        for (key, storage_val) in aggregator_keys.into_iter() {
+            let mut latest_value: Option<u128> = match storage_val
+                .ok() // Was anything found in storage
+                .map(|value| value.map(|bytes| deserialize(&bytes)))
+            {
+                None => None,
+                Some(v) => v,
+            };
+
+            let indexed_entries = self
+                .versioned_outputs
+                .indexed_entries(&key)
+                .expect("No entries found for the provided key");
+            for (idx, entry) in indexed_entries.iter() {
+                match &entry.cell {
+                    EntryCell::Write(_, data) => {
```

Comment:
If this is a delete, then we will later panic. We need to document that we currently assume no deletion and deal with this later before enabling other use cases.

After discussion with Zekun, we are probably fine because we will have a read-write conflict accessing the aggregator. Need to test.

Reply:
I was discussing with @gelash : deletion is a tricky case since in Move we would consume a value, which means that the resource that holds aggregator (which we delete) must be deleted as well, right? Examples:
```
tx2: remove aggregator from table and remove it
tx3: borrow aggregator from table and add // fails on borrow!
```
I think we have the same thing with vector and option (implemented as vector under the hood)

Reply:
Or better say: whatever holds aggregator controls access to it. if we remove aggregator, we fail on accessing it via that holding resource, or fail when resource is also destroyed

Reply:
yeah I think the deletion and delta op will be detected as write/read conflict on the container of the resource itself, but we can't panic because of the speculative execution